workshop.pl

```diff
@@ -541,35 +547,46 @@ sub arg_handler {
     logger('info', "'$req'...\n", 1);
 
     my $tmp_req = {};
-
     logger('info', "importing JSON...\n", 2);
-    if (open(my $req_fh, "<", $req)) {
-        my $file_contents;
-        while(<$req_fh>) {
-            $file_contents .= $_;
-        }
+    (my $rc, $tmp_req->{'json'}) = get_json_file($req, $schema_location);
 
+    if ($rc == 0 and defined $tmp_req->{'json'}) {
         $tmp_req->{'filename'} = $req;
-        $tmp_req->{'json'} = decode_json($file_contents);
-
-        close($req_fh);
-
         logger('info', "succeeded\n", 3);
     } else {
         logger('info', "failed\n", 3);
-        logger('error', "Failed to load requirement file '$req'!\n");
-        exit(get_exit_code('failed_opening_requirement'));
-    }
-
-    logger('info', "validating JSON...\n", 2);
-    my @json_v_errors = $json_v->validate($tmp_req->{'json'});
-    if (scalar(@json_v_errors) > 0) {
-        logger('info', "failed\n", 3);
-        logger('error', "Could not JSON validate the requirement file '$req'!\n");
-        logger('error', Dumper(\@json_v_errors));
-        exit(get_exit_code('requirement_failed_validation'));
-    } else {
-        logger('info', "succeeded\n", 3);
+        #logger('error', "Failed to load requirement file '$req'!\n");
+        if ($rc == 1) {
+            logger('error', "Schema file undefined!\n");
+            exit(get_exit_code('no_schema'));
+        } elsif ($rc == 2) {
+            logger('error', "Schema file '$schema_location' not found!\n");
+            exit(get_exit_code('schema_not_found'));
+        } elsif ($rc == 3) {
+            logger('error', "Cannot open schema file: '$schema_location'\n");
+            exit(get_exit_code('failed_opening_schema'));
+        } elsif ($rc == 4) {
+            logger('error', "Schema '$schema_location' is invalid JSON!\n");
+            exit(get_exit_code('schema_invalid_json'));
+        } elsif ($rc == 5) {
+            logger('error', "Schema validation for '$req' using schema '$schema_location' failed!\n");
+            exit(get_exit_code('requirement_failed_validation'));
+        } elsif ($rc == 6) {
+            logger('error', "Requirement '$req' is invalid JSON!\n");
+            exit(get_exit_code('requirement_invalid_json'));
+        } elsif ($rc == 7) {
+            logger('error', "Requirement file name undefined\n");
+            exit(get_exit_code('no_requirement'));
```

Comment:
How do we attempt to load a requirement if there is no requirement?

Reply:
You shouldn't load a requirement if there is no req file defined, but this is a valid return code for get_json_file, assuming the caller provided a var for the file that was empty.  If you already handle checking for that case earlier, then we can skip it here, but I thought it would be goo practice to handle every possible return code get_json_file can provide.

Reply:
well, we already ignore many return codes in many places all the time...it just seems extraneous to be checking for ones that cannot happen, IMHO